monai/metrics/metric.py

```diff
@@ -0,0 +1,84 @@
+# Copyright 2020 - 2021 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from typing import Any, List, Optional, Sequence, Union
+
+import torch
+
+from monai.config import TensorOrList
+
+
+class Metric(ABC):
+    """
+    Base class of Metrics interface.
+    `__call__` is supposed to compute independent logic for several samples of `y_pred` and `y`(optional).
+    Ususally, subclass only needs to implement the `_apply` function for computation process.
+    And `reduce` is supposed to execute reduction for the final result, it can be used for 1 batch data
+    or for the accumulated overall data.
+
+    """
+
+    def __call__(self, y_pred: TensorOrList, y: Optional[TensorOrList] = None):
+        """
+        Execute basic computation for model prediction and ground truth.
+        It can support  both `list of channel-first Tensor` and `batch-first Tensor`.
+        And users can execute on every batch of data, then accumulate the results, or
+        accumulate the original `y_pred` and `y`, then execute on the accumulated data.
+
+        Args:
+            y_pred: the model prediction data to compute, must be a list of `channel-first` Tensor
+                or a `batch-first` Tensor.
+            y: the ground truth to compute, must be a list of `channel-first` Tensor
+                or a `batch-first` Tensor.
+
+        """
+        ret: Union[torch.Tensor, Sequence[torch.Tensor]]
+        if isinstance(y_pred, (list, tuple)) or isinstance(y, (list, tuple)):
+            # if y_pred or y is a list of channel-first data, add batch dim and compute metric
+            ret_: List[torch.Tensor]
+            if y is not None:
+                ret_ = [self._apply(p_.detach().unsqueeze(0), y_.detach().unsqueeze(0)) for p_, y_ in zip(y_pred, y)]
```

Comment:
a few questions about the base class:
- how to support multiprocessing `_apply` in parallel? this class shouldn't provide the concrete implementation of the list processing, for example, we may have subclasses to do:
```
with multiprocess.pool as p:
  p.map(self._apply, zip(y_pred, y))
```

- `_apply` is unclear to me, perhaps `_compute` is better?

- I feel `reduce` could be `aggregate`, reduce is a parallel computing concept, it doesn't suit all our use cases such as computing ROC AUC.

- the postprocesses after line 52 including `ret = torch.cat(ret_, dim=0)` look like a kind of reduce/aggregate as well, why not implement them in `self.reduce`?

- does this abstraction support the pathology froc metric? https://github.com/Project-MONAI/MONAI/blob/dev/monai/apps/pathology/metrics.py#L34

Reply:
* Reduction needs to be done across processes in a multi-GPU situation as well, see [Ignite's facility](https://github.com/pytorch/ignite/blob/dce97dfeacf1516e78b1cd9c6a37e57a1228ceed/ignite/metrics/metric.py#L519) for doing this. 
* `_compute` does sound better. 
* `aggregate` is a better term, I had thought `reduce` was a parallel operation

Reply:
Hi @ericspod and @wyli ,

Thanks very much for your comments and suggestions, about the questions:
1. I added `_compute_list()` protected method, so a subclass can easily re-implement it with multi-threads.
2. Changed `_apply()` to `_compute()`.
3. Changed `reduce()` to `aggregate()` and updated all the tests.
4. I added this concat logic in `call()` is to make the output of the function to be a `batch-first` Tensor for both `list of channel-first tensor` and `batch-first tensor` inputs. I think we should not return `list of batch-first tensor` which is not a supported data shape in our design. Or maybe we can split this post-processing logic into a separate function?
5. I checked the FROC metric, seems it computes on some numpy array data in files which was computed by `y_pred` and `y` before. So the computation is actually only the `aggregation` part. Once we finalized the `Metric` base class, I can update it if necessary.

What do you think?

Thanks.

Reply:
Hi @ericspod @wyli @rijobro , do you have any more comments on this PR?

Thanks in advance.